Add create project form to web console

spadgett opened the pull request:
@jwforres @liggitt WIP - thoughts?

![create-project-button](https://cloud.githubusercontent.com/assets/1167259/7371825/432177e6-ed91-11e4-9849-200cd7ca4e69.png)

![create-project](https://cloud.githubusercontent.com/assets/1167259/7371829/46b395d8-ed91-11e4-946b-1ea4b13d8e74.png)

![create-project-invalid](https://cloud.githubusercontent.com/assets/1167259/7371832/4b476354-ed91-11e4-9d84-05582cb1673b.png)

jwforres commented:
asterisk to indicate a required field, name is the only thing required

jwforres commented:
might help to explain why name and display name are both there, maybe just a help tip next to name that says it is the unique name for the project.

spadgett commented:
I fixed an issue related to a rebase and updated the UI. @jwforres is this what you meant?

![name-updates](https://cloud.githubusercontent.com/assets/1167259/7373003/fbcfdef8-ed97-11e4-9707-b9cbdfe7afa1.png)

jwforres commented:
@spadgett yes on the asterisk.  Wasn't what i originally intended on the help, but that works too.

spadgett commented:
@jwforres @liggitt please review. I squashed the changes to one commit.

I have the change to use the API from #1948 implemented locally... I was planning on submitting a new PR when #1948 is merged. (My local change also removes the "osc new-project" message, but only when the Create Project button is there.)

spadgett commented:
@liggitt I have all the changes in now, and #1948 is merged. Please review.

spadgett commented:
@liggitt FYI, I've rebased from upstream master

spadgett commented:
@liggitt I think I've fixed the Travis failure.

spadgett commented:
@liggitt Whenever you have to time look. Changes rebased again.

liggitt commented on assets/app/scripts/controllers/createProject.js:
```diff
@@ -0,0 +1,34 @@
+  .controller('CreateProjectController', function ($scope, DataService, Notification, Navigate) {
+    $scope.createProject = function() {
+      if($scope.createProjectForm.$valid) {
+        var p = DataService.create('projectRequests', null, {
```
does `var p` do anything?

liggitt commented on assets/app/scripts/controllers/createProject.js:
```diff
@@ -0,0 +1,34 @@
+          annotations: {
+            description: $scope.description
+          }
+        }, $scope)
```
`.then` on a newline makes my nervous about JS parsing...

liggitt commented on assets/app/scripts/controllers/projects.js:
```diff
@@ -19,4 +20,32 @@ angular.module('openshiftConsole')
+        if (result.status !== 0) {
+          msg += " (" + result.status + ")";
+        }
+        Notification.error(msg);
```
not sure I'd flaunt this... maybe log it as a warning?

spadgett commented:
@liggitt updated for your comments

spadgett commented:
@liggitt Is there anything you were waiting on from me on this PR? I see the travis build failed, but it didn't look related to my change.

smarterclayton commented:
[test]

spadgett commented:
@liggitt spurious?

```
!!! Error in hack/test-cmd.sh:301
  'registrypod=$(osc get pod | grep docker-registry-1 | awk '{print $1}')' exited with status 1
Call stack:
  1: hack/test-cmd.sh:301 main(...)
Exiting with status 1
[FAIL] !!!!! Test Failed !!!!
2015/05/05 23:44:25 received terminated signal, shutting down...
make: *** [check-test] Error 1
The command "PATH=$HOME/gopath/bin:./_output/etcd/bin:$PATH make check-test WHAT="''" TESTFLAGS="-p=4"" exited with 2.
```

smarterclayton commented:
That was fixed a while ago - rebase maybe?

----- Original Message -----
> @liggitt spurious?
> 
> ```
> !!! Error in hack/test-cmd.sh:301
>   'registrypod=$(osc get pod | grep docker-registry-1 | awk '{print $1}')'
>   exited with status 1
> Call stack:
>   1: hack/test-cmd.sh:301 main(...)
> Exiting with status 1
> [FAIL] !!!!! Test Failed !!!!
> 2015/05/05 23:44:25 received terminated signal, shutting down...
> make: *** [check-test] Error 1
> The command "PATH=$HOME/gopath/bin:./_output/etcd/bin:$PATH make check-test
> WHAT="''" TESTFLAGS="-p=4"" exited with 2.
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1949#issuecomment-101394504

spadgett commented:
Rebased....

liggitt commented on assets/app/scripts/services/data.js:
```diff
@@ -825,7 +826,7 @@ angular.module('openshiftConsole')
       deferred.resolve({namespace: opts.namespace});
     }
-    else if (context.projectPromise && type !== "projects") {
+    else if (context.projectPromise && type !== "projects" && type !== "projectRequests") {
```
is this necessary?

spadgett replied:
@liggitt You're right. I'll remove it.

liggitt commented on assets/app/views/createProject.html:
```diff
@@ -0,0 +1,66 @@
+                placeholder="my-project"
+                type="text" autofocus
+                required
+                maxlength="63"
```
go ahead and add minlength="2" (c.f. https://github.com/openshift/origin/issues/1232#issuecomment-100232543)

liggitt replied:
actually... let's not be more restrictive than the API

liggitt commented:
Opened a PR with some tweaks: https://github.com/spadgett/origin/pull/2

liggitt commented:
After creating a project, the projects selector doesn't contain the project I just created, so the currently selected item is wrong

spadgett commented:
@liggitt We're watching projects in nav.js. The callback is invoked a few times, but it doesn't have the new project. Is this a bug?

```
        $scope.$watch("projects", function(projects) {
          select.empty();
          updateOptions(projects);
          select.selectpicker('refresh');
        });
```

The project is definitely created.

liggitt commented on assets/app/views/projects.html:
```diff
@@ -1,18 +1,25 @@
-  <div style="margin-top: 10px;">
-    To create a new project, run <code>osc new-project &lt;projectname&gt;</code>
+  <div ng-if="canCreate === false" style="margin-top: 10px;">
+    <span ng-if="!newProjectMessage">To create a new project, run <code>osadm new-project &lt;projectname&gt;</code></span>
```
need to add `--admin=...` to this example command

liggitt commented on assets/app/scripts/controllers/projects.js:
```diff
@@ -19,4 +20,37 @@ angular.module('openshiftConsole')
+      }
+
+      // Check if there are detailed messages. If there are, show them instead of our default message.
+      if (data.details) {
```
@deads2k moving discussion about message here. there are three cases where we get 403:
1. no access to list projectRequests. the returned error message isn't suitable to display
2. access to list projectRequests, but no access to create projects. the default message isn't good to display today... we could probably come up with a better default.
3.  no access with a custom message set.

I could see coming up with a better message for 2, but the trick is distinguishing between those and 1. Putting the message we wanted the UI to surface in details was the quickest way I could think of to signal that. Open to other ideas.

deads2k replied:
> Putting the message we wanted the UI to surface in details was the quickest way I could think of to signal that.

It's ugly, but its internal.  I'll live with it.

liggitt commented on assets/app/scripts/directives/nav.js:
```diff
@@ -27,7 +27,27 @@ angular.module('openshiftConsole')
-        var updateOptions = function(projects) {
+        var updateOptions = function() {
+          // Wait until everything we need is set.
+          if (!$scope.projects ||
```
actually, if we have `$scope.project` (which we will), we want to show that in the dropdown while we're waiting for the projects list to come back, instead of "No value"

liggitt commented on assets/app/scripts/directives/nav.js:
```diff
@@ -27,7 +27,28 @@ angular.module('openshiftConsole')
-        var updateOptions = function(projects) {
+        var updateOptions = function() {
+          var currentProject = $scope.project;
+          if (!currentProject || angular.equals({}, currentProject)) {
```
also make sure we have `$scope.projectName` before constructing, just in case that doesn't exist

liggitt commented on assets/app/scripts/directives/nav.js:
```diff
@@ -27,7 +27,28 @@ angular.module('openshiftConsole')
+          // doesn't exist. This can happen when creating a new project from
+          // the web, which navigates immediately to the project page.
+          var projects;
+          if (!$scope.projects || !$scope.projects[$scope.projectName]) {
```
same here... only if `$scope.projectName` exists

liggitt commented on assets/app/scripts/directives/nav.js:
```diff
@@ -57,11 +78,14 @@ angular.module('openshiftConsole')
         });
-        $scope.$watch("projects", function(projects) {
+
+        var clearAndUpdateOptions = function() {
```
`$scope.$watch` calls this immediately... according to https://docs.angularjs.org/api/ng/type/$rootScope.Scope#$watch you can detect and ignore this case by doing an `===` comparison on the `newValue`,`oldValue` args and returning early when true. Should probably do that or the selector will be rebuilt three times every time we navigate... I'm sure that doesn't help the flickering.

liggitt commented on assets/app/views/createProject.html:
```diff
@@ -0,0 +1,66 @@
+              ng-click="createProject()"
+              ng-disabled="createProjectForm.$invalid || nameTaken"
+              value="">Create</button>
+          <a class="btn btn-default btn-lg" href="#" back>Cancel</a>
```
where does this take you if this is the first page you've loaded? I'd almost expect this to consistently take you do the projects page

liggitt commented on assets/app/views/projects.html:
```diff
@@ -1,18 +1,28 @@
+    <span ng-if="!newProjectMessage">An administrator can create a project for you
+      by running the command
+      <code>osadm new-project &lt;projectname&gt;
+        --admin={{user.metadata.name || '&lt;YourUsername&gt;'}}</code></span>
```
pretty aggressive wrapping :)
```
<span ng-if="!newProjectMessage">
  An administrator can create a project for you by running the command
  <code>osadm new-project &lt;projectname&gt; --admin={{user.metadata.name || '&lt;YourUsername&gt;'}}</code>
</span>
```
easier to read as this?

spadgett replied:
No :)

![2__vim__vim_](https://cloud.githubusercontent.com/assets/1167259/7659641/763bb88e-fb11-11e4-8576-c0fc4596453c.png)

But I'll change it.

liggitt commented:
[merge]

liggitt commented:
[merge]
